Minesweeper 게임 리팩토링 및 기능 추가

1seyoung opened the pull request:
- 구현 내용 :

> **게임 보드 생성 리팩토링**
>기존 2차원 배열에 단순 데이터를 저장하던 방식을 개선 , Tile 객체를 활용하여 타일 상태(지뢰 여부, 깃발 여부, 주변 지뢰 개수 등)를 개별적으로 관리 수행
>
> **승리 상태 판단 로직 구현**
>  GameBoard.java 에 allSafeOpen 메서드를 추가하여 승리 조건 확인
>  승리 조건이 충족되면 **GameState**를 VICTORY로 변경
>
> **상태 기반 시스템 구성**
> 게임 상태를 기반으로 동작하도록 리팩토링
> GameState.java: 게임 상태를 나타내는 Enum(INITIAL, PLAYING, VICTORY, GAME_OVER)
> GameStateModel.java: 현재 게임 상태를 저장하고 이벤트 기반으로 상태를 전환하는 로직 추가.
>
> **컨트롤러 구현**
> 사용자 명령을 게임 보드와 상태 관리 로직에 전달
> handleTileClick(x, y): open 명령 처리 , 지뢰면 GAME_OVER 상태로 전환, 안전하면 해당 타일 열고 오픈
> handleTileFlag(x, y): flag 명령 처리 
 ``` com. practice. GameBoardpublic void checkedFlag( int x, int y )``` 


- 메인 리뷰어 지정 및 Due Date
@blue000927 

- 고민 포인트, 리뷰 시 참고 사항(코드)
상태 기반으로 시스템이 동작할 수 있도록 코드 구현
이벤트 핸들러를 따로 두는게 좋은지 고민(현재 GameStateModel 에서 상태 및 이벤트 모두 관리)

- 체크리스트
    - [x] PR 제목을 명령형으로 작성했습니다.
    - [ ] PR을 연관되는 github issue에 연결했습니다.
    - [x] 리뷰 리퀘스트 전에 셀프 리뷰를 진행했습니다
    - [ ] 변경사항에 대한 테스트코드를 추가했습니다. 또는, 테스트 코드가 필요없는 이유가 있습니다

1seyoung committed:
refactor branch test

1seyoung committed:
refactor

1seyoung committed:
refactor

blue000927 commented on .idea/gradle.xml:
.gitignore 파일을 통해 .idea는 git에 트래킹되지 않도록 설정해주시면 좋을듯합니다!

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,193 @@
+package com.practice;
+
+
+
+public class GameBoard {
```
패키지와 클래스 사이에는 \n 하나만 씁니다.

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,193 @@
+        this.board = new Tile[rows][cols];
+        this.numMines = 0;
+        initializeBoard(rows, cols, level);
+
+
```
불필요한 \n 제거

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,193 @@
+        //okay
+        placeMines(numMines); // 여기서 실패(12.23.7시 02분) -> 해결
+        countSurroundMine();
+
+
+
```
전체적으로 이런 불필요한 \n들은 체크해서 지워주세요,

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,193 @@
+
+        numMines = (int) (rows * cols * _level.getMineRatio());
+
+        // 기존 Game.java 85 - 105 line
```
주석은 정말 필요한 상황에만 작성합니다!
주석 대신 코드만으로도 이해할 수 있도록 변수명, 함수명, 로직의 구성 등을 신경써서 작성해주세요. 일단 모든 주석을 제거합시다.

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,193 @@
+
+    private void initializeBoard(int rows, int cols, String level) {
+
+        GameLevel _level = GameLevel.fromInput(level);
```
'_'를 왜 붙이신건가요?

1seyoung replied:
'_'는 python 으로 코딩하던 예전 버릇이고 여기서는 `level`과 구분하려는 의도로 사용했습니다. 변수명 gameLevel 로 수정하였습니다.

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,193 @@
+                }
+            }
+            System.out.println();
+        }
```
MVC 패턴에 대해 학습해보시고 요런 로직은 View 로직을 담당하는 객체가 가져가 보면 좋겠습니다.

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,193 @@
+
+                // 인접 타일이 열려 있지 않고, 지뢰가 아닌 경우만 탐색
+                if (!neighborTile.isOpened() && !neighborTile.isMine()) {
+                    openTile(nx, ny, gameState); // 재귀적으로 인접 타일 열기
```
1. 재귀 방식의 장단점에 대해 설명해주세요.
2. 재귀가 아닌 다른 방식으로도 구현할 수 있을까요?

blue000927 commented on src/main/java/com/practice/GameController.java:
```diff
@@ -0,0 +1,39 @@
+
+public class GameController {
+    // 게임 상태, 이벤트 , 게임 보드 관리하는 컨트롤러
+    private GameStateModel gameState; // 왜 final 로 하라는 거야...
+    private GameBoard gameBoard;  // 여기가 public 이고 아니고가 영향을 주는감...
```
이런 궁금증은 주석으로 남기시지말고 PR에서 코멘트를 통해 저에게 질문해주세요.
혹은 별도 문서에서 고민해주시면 좋겠습니다. (깃허브 커밋은 모두 기록에 남습니다.)

blue000927 commented on src/main/java/com/practice/GameController.java:
```diff
@@ -0,0 +1,39 @@
+package com.practice;
+
+public class GameController {
```
컨트롤러가 하는 역할은 무엇인가요?

blue000927 commented on src/main/java/com/practice/GameLevel.java:
```diff
@@ -0,0 +1,34 @@
+    // 입력 값에 따라 난이도 반환 (유효하지 않은 경우 EASY 반환)
+    public static GameLevel fromInput(String input) {
+        switch (input.toUpperCase()) {
+            case "EASY":
```
EASY라는 키워드는 이미 enum 열거형 상수 중에 EASY가 있는데 이를 이용할 수 없을까요?

blue000927 commented on src/main/java/com/practice/GameStateModel.java:
```diff
@@ -0,0 +1,55 @@
+    public void handleEvent(String event) {
+        switch (currentState) {
+            case INITIAL:
+                if (event.equals("game_start")) {
```
game_start라는 값을 상수로 분리할 수는 없을까요? 혹은 enum으로

blue000927 reviewed (commented):
추가로 Junit5 + AssertJ를 통해 각 클래스에 대해 단위 테스트를 작성해봐 주시면 감사하겠습니다.

1seyoung committed:
Update .gitignore and untrack files

1seyoung committed:
이벤트 문자열 비교 방식을 enum으로 변경
- GameEvent enum 추가: 이벤트를 문자열 대신 enum으로 관리
- handleEvent 메서드에서 문자열 비교 제거

상태 인터페이스 구현
- 기존 스위치를 사용한 상태 전환에서 이벤트 값에 따라 현재 상태 객체가 변경되도록 수정
- 각 상태 객체가 자신의 동작을 처리하도록 책임 분리

GameView 구현
- 보드 출력 관련 코드를 별도 View 클래스로 분리

etc
- 주석 제거 등 간단한 수정

1seyoung commented:
피드백 반영 및 상태,이벤트 로직 개선 7c3ee766d244c36d8150e5f96a402f8f26a8e2aa

**구현 내용**
> 이벤트 문자열 비교 방식을 enum으로 변경
> - GameEvent enum 추가: 이벤트를 문자열 대신 enum으로 관리
> - handleEvent 메서드에서 문자열 비교 제거하여 코드 가독성 및 타입 안정성 향상
> 
> 상태 인터페이스 구현
> - 기존 스위치를 사용한 상태 전환 로직을 개선
> - 이벤트 값에 따라 현재 상태 객체가 변경되도록 수정
> - 각 상태 객체가 자신의 동작을 처리하도록 책임 분리하여 유지보수성 강화
> 
> 보드 출력 로직을 View로 분리
> - 보드 출력 관련 코드를 별도 View 클래스로 분리
> - Main 클래스의 책임을 줄이고 가독성 개선
> 
> etc
> - 불필요한 주석 제거 및 간단한 코드 정리


고민 포인트, 리뷰 시 참고 사항(코드)
비어있는 블록을 클릭할 시 주변 블록을 open 하는 알고리즘으로 재귀말고 어떤 것으로 구현하면 좋을지 잘 모르겠습니다.

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,155 @@
+            }
+        }
+        if (allSafeOpen()) {
+            //gameState.handleEvent("all_safe");
```
주석은 제거해주세요!

blue000927 commented on src/main/java/com/practice/GameController.java:
```diff
@@ -0,0 +1,39 @@
+
+    public void handleTileClick(int x, int y) {
+        Tile tile = gameBoard.getTile(x,y);
+        //gameState.handleEvent("click_tile");
```
주석은 제거해주세요!(ditto : 이전 리뷰와 동일하다는뜻)

blue000927 commented on src/main/java/com/practice/GameController.java:
```diff
@@ -0,0 +1,39 @@
+        gameState.handleEvent(GameEvent.CLICK_TILE);
+
+        if (tile.isMine()) {
+            //gameState.handleEvent("click_mine");
```
ditto

blue000927 commented on src/main/java/com/practice/GameController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public void startGame() {
+        //gameState.handleEvent("game_start");
```
ditto

blue000927 commented on src/main/java/com/practice/GameLevel.java:
```diff
@@ -0,0 +1,26 @@
+        return mineRatio;
+    }
+
+    public static GameLevel fromInput(String input) {
+        try {
+            return GameLevel.valueOf(input.toUpperCase());
+        } catch (IllegalArgumentException e) {
+            return EASY;
+        }
+    }
```
이렇게 예외가 발생하면 무조건 EASY로 설정했을 때의 문제점은 없을까요?

blue000927 commented on src/main/java/com/practice/GameBoard.java:
```diff
@@ -0,0 +1,155 @@
+package com.practice;
+
+public class GameBoard {
```
패키지를 model, view, controller로 분리하여 각각 해당하는 패키지에 클래스를 위치시켜주시면 감사하겠습니다.

blue000927 commented on src/main/java/com/practice/state/GameState.java:
```diff
@@ -0,0 +1,9 @@
+import com.practice.GameEvent;
+import com.practice.GameStateModel;
+
+public interface GameState {
```
오호 상태 패턴을 구현하셨군요!
상태 패턴을 적용하신 이유와 기존 코드 대비 더 좋아진 점에 대해 설명해주시겠어요?

blue000927 commented on src/main/java/com/practice/state/InitialState.java:
```diff
@@ -0,0 +1,27 @@
+    public void handleEvent(GameEvent event, GameStateModel model) {
+        if (event == GameEvent.GAME_START) {
+            model.setCurrentState(ReadyState.getInstance());
+            System.out.println("Game started! State is now READY.");
```
InitialState에서 READY를 출력하는 것이 좀 부자연스러워보이네요.

blue000927 commented on src/main/java/com/practice/state/ReadyState.java:
```diff
@@ -0,0 +1,29 @@
+    public void handleEvent(GameEvent event, GameStateModel model) {
+        if (event == GameEvent.CLICK_TILE) {
+            model.setCurrentState(PlayingState.getInstance());
+            System.out.println("Game is now in PLAYING state.");
```
여기도 READY면 ready 관련 출력을 해야 하는 것 아닌가요?

blue000927 commented on src/main/java/com/practice/state/GameOverState.java:
```diff
@@ -0,0 +1,24 @@
+        return instance;
+    }
+
+    @Override
+    public void handleEvent(GameEvent event, GameStateModel model) {
+        System.out.println("Game is already over. Restart to play again.");
+    }
```
GameOverState는 mvc 중에 무엇인가요?
model에 더 가까워보이는데 print 같은 출력문이 끼어서 한 객체가 2가지 역할을 갖고 있는 것으로 보여요.

blue000927 commented on src/main/java/com/practice/Main.java:
```diff
@@ -1,63 +1,72 @@
-                    System.exit(0); // 시스템 종료
+        Scanner scanner = new Scanner(System.in);
+
+        System.out.println("Enter rows: ");
+        int rows = scanner.nextInt();
+
+        System.out.println("Enter columns: ");
+        int cols = scanner.nextInt();
```
InputView, OutputView 객체를 추가하여 인풋과 아웃풋 담당을 해 보면 어떨까요?

blue000927 commented on src/main/java/com/practice/Main.java:
```diff
@@ -1,63 +1,72 @@
+        }
+
+        if (command.equals("start")) {
+            gameController.startGame();
+            gameController.printBoard();
+            while (true) {
+                GameState currentState = gameState.getCurrentState();
```
controller 객체와 main 객체는 어떠한 차이가 있나요?
